### PR TITLE
simplifier pass: create temporary for RHS of an assignment (only ArrayConstructor for now)

### PR DESCRIPTION
**Pull request by gxyd**
## What changed?

The idea with the PR is simple, for an assignment the expression on RHS never needs a temporary variable, except when:
* realloc-lhs is true
* LHS is an allocatable
* The Var on LHS is present on the RHS

with these conditions being true, we need to create a temporary variable for RHS. Currently as there are limited test cases for it, I only handled the case for when RHS is an ArrayConstructor of arrays as elements in it.

## For later
It's possible to have ArrayConstructor of StructType as well, we can probably handle that later, as currently there is no test case for it either.

### CODE REVIEW

File: src/libasr/pass/simplifier.cpp [hunk 1]
```diff
@@ -1002,6 +1004,41 @@ class ArgSimplifier: public ASR::CallReplacerOnExpressionsVisitor<ArgSimplifier>
...
+                ASR::Var_t* v1 = ASR::down_cast<ASR::Var_t>(xx.m_target);
+                bool create_temp_var_for_rhs = false;
+                Vec<ASR::expr_t*> array_vars; array_vars.reserve(al, 1);
+                ArrayVarCollector array_var_collector(al, array_vars);
+                array_var_collector.visit_expr(*xx.m_value);
+                // after collecting variables from RHS, we check whether
+                // there is any common variable
+                for (size_t i=0; i < array_vars.size(); i++) {
+                    ASR::Var_t* v = ASR::down_cast<ASR::Var_t>(array_vars[i]);
+                    if (v->m_v == v1->m_v) {
```
Comment (gxyd) [inline src/libasr/pass/simplifier.cpp:1027]: I'm not sure if this is the best way to do, what I intend to do here.

**Comment by gxyd** [comment]
I'll move it to "Ready for review" as I intend to received feedback on this.

**Comment by certik** [comment]
I think this is fine, especially since it only triggers for `reallocate_lhs`.

Can you show a Fortran example where this is needed and show what it is doing? I can provide more feedback then.

**Comment by gxyd** [comment]
> I think this is fine, especially since it only triggers for `reallocate_lhs`.
> 
> Can you show a Fortran example where this is needed and show what it is doing? I can provide more feedback then.

Yes, below is the MRE this PR intends to fix (for `arrays_op_18.f90`):
```f90
program arrays_op_18
   implicit none

   real, allocatable :: a(:)
   real :: b(2, 4)
   allocate(a(2))

   a = 5.0
   b = 6.0

   a = [b, a]

   print *, lbound(a, 1), ubound(a, 1)

end program
```

currently on *simplifier pass* branch, on execution with `--realloc-lhs` flag, this outputs:
```
$ lfortran test.f90 --realloc-lhs
1    2
```

while on this branch:
```
$ lfortran test.f90 --realloc-lhs
1    10
```

which is correct.


The main difference being that with the output of *simplifier pass* is different now:
```diff
> real(4), dimension(:), allocatable :: __libasr_created__assignment_
9,11c10,13
< deallocate(a)
< allocate(a(2))
< a = [b, a]
---
> deallocate(__libasr_created__assignment_)
> allocate(__libasr_created__assignment_(size(a) + 8))
> __libasr_created__assignment_ = [b, a]
> a = __libasr_created__assignment_
```

as can be seen above, we create a temporary for RHS (i.e. `[b, a]`), the rest of things are then easily handled by *array op* pass.

**Comment by certik** [comment]
I see. It looks good to me. We can further iterate on such things once simplifier is merged.

**Pull request merged**